lms/resources/lti_launch.py

```diff
@@ -30,10 +30,17 @@ def __init__(self, request):
     def course(self):
         """Get the course this LTI launch based on the request's params."""
 
+        extra = self._new_course_extra()
+        if existing_course := self._request.find_service(
+            name="course"
+        ).get_by_context_id(self._request.parsed_params["context_id"]):
+            # Keep existing `extra` instead of replacing it with the default
+            extra = existing_course.extra
+
         return self._request.find_service(name="course").upsert_course(
             context_id=self._request.parsed_params["context_id"],
             name=self._request.parsed_params["context_title"],
-            extra=self._course_extra(),
+            extra=extra,
```

Comment:
We are using `self._request.find_service(name="course")` twice here, we should probably keep a reference.

On the other hand, this is the only time we ever call this method in the codebase. If we need to make another call to the service immediately before, this strongly suggests that this should get pushed down into the service.

Reply:
> this strongly suggests that this should get pushed down into the service.

Yeah, that would be much nicer, but `new_course_extra` needs a `is_canvas`. I have another refactor going on to disentangle  LTIParams and Product and that will go nicely with this.

It would also move more stuff out of `resource/lit_launch.py` which an odd spot in our code base now.


I'll leave as is now so it's more obvious what need to be done later. Made a note of this :+1: